// cli-engine-heroku/lib/command.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _cliEngineCommand = require('cli-engine-command');

var _cliEngineCommand2 = _interopRequireDefault(_cliEngineCommand);

var _api_client = require('./api_client');

var _api_client2 = _interopRequireDefault(_api_client);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class HerokuCommand extends _cliEngineCommand2.default {

  async init() {
    await super.init();
    if (this.flags.app) this.app = this.flags.app;
  }

  get heroku() {
    if (this._heroku) return this._heroku;
    this._heroku = new _api_client2.default(this);
    return this._heroku;
  }
  get legacyHerokuClient() {
    if (this._legacyHerokuClient) return this._legacyHerokuClient;
    const HerokuClient = require('heroku-client');
    let options = {
      // flow$ignore
      userAgent: this.heroku.defaultOptions.headers['user-agent'],
      debug: this.config.debug,
      token: this.heroku.auth,
      host: `${this.heroku.defaultOptions.protocol || 'https:'}//${this.heroku.defaultOptions.host || 'api.heroku.com'}`
    };

    this._legacyHerokuClient = new HerokuClient(options);
    return this._legacyHerokuClient;
  }
}
exports.default = HerokuCommand;
