// cli-engine-command/lib/command.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _output = require('./output');

var _output2 = _interopRequireDefault(_output);

var _parser = require('./parser');

var _parser2 = _interopRequireDefault(_parser);

var _package = require('../package.json');

var _package2 = _interopRequireDefault(_package);

var _cliEngineConfig = require('cli-engine-config');

var _httpCall = require('http-call');

var _httpCall2 = _interopRequireDefault(_httpCall);

var _help = require('./help');

var _help2 = _interopRequireDefault(_help);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-line
class Command {

  static get id() {
    let cmd = this.command ? `${this.topic}:${this.command}` : this.topic;
    return cmd;
  }

  /**
   * instantiate and run the command setting {mock: true} in the config (shorthand method)
   */
  static async mock(...argv) {
    return this.run({ argv, mock: true });
  }

  /**
   * instantiate and run the command
   */
  static async run(options = {}) {
    let config = (0, _cliEngineConfig.buildConfig)(options.config);
    let output = options.output || new _output2.default({ mock: options.mock, config });
    let cmd = new this({ config, output });
    cmd.argv = options.argv || [];
    // if (this.flags.debug) this.config.debug = 1
    try {
      const args = await cmd.init();
      await cmd.run(args);
      await cmd.out.done();
    } catch (err) {
      cmd.out.error(err);
    }
    return cmd;
  }

  constructor(options = {}) {
    this.flags = {};
    this.args = {};

    this.config = (0, _cliEngineConfig.buildConfig)(options.config);
    this.out = options.output || new _output2.default({ config: this.config });
    this.http = _httpCall2.default.defaults({
      headers: {
        'user-agent': `${this.config.name}/${this.config.version} (${this.config.platform}-${this.config.arch}) node-${process.version}`
      }
    });
  }

  async init() {
    const parser = new _parser2.default({
      flags: this.constructor.flags || {},
      args: this.constructor.args || [],
      variableArgs: this.constructor.variableArgs,
      cmd: this
    });
    const { argv, flags, args } = await parser.parse({ flags: this.flags, argv: this.argv });
    this.flags = flags;
    this.argv = argv;
    this.args = args;
  }

  // prevent setting things that need to be static


  /**
   * actual command run code goes here
   */
  async run(...rest) {}

  get stdout() {
    return this.out.stdout.output;
  }

  get stderr() {
    return this.out.stderr.output;
  }

  static buildHelp(config) {
    let help = new _help2.default(config);
    return help.command(this);
  }

  static buildHelpLine(config) {
    let help = new _help2.default(config);
    return help.commandLine(this);
  }
}
exports.default = Command;
Command.aliases = [];
Command.variableArgs = false;
Command.args = [];
