// cli-ux/lib/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Config {
    static get debug() {
        return !!Config._globals.debug;
    }
    static set debug(debug) {
        Config._globals.debug = debug;
    }
    static get mock() {
        return !!Config._globals.mock;
    }
    static set mock(mock) {
        if (mock) {
            Config.stdout = '';
            Config.stderr = '';
        }
        Config._globals.mock = mock;
    }
    static get displayTimestamps() {
        return !!Config._globals.displayTimestamps;
    }
    static set displayTimestamps(displayTimestamps) {
        Config._globals.displayTimestamps = displayTimestamps;
    }
    static get errlog() {
        return Config._globals.errlog;
    }
    static set errlog(errlog) {
        Config._globals.errlog = errlog;
    }
    static get stdout() {
        return Config._globals.stdout || '';
    }
    static set stdout(stdout) {
        Config._globals.stdout = stdout;
    }
    static get stderr() {
        return Config._globals.stderr || '';
    }
    static set stderr(stderr) {
        Config._globals.stderr = stderr;
    }
    static get windows() {
        return process.platform === 'win32';
    }
    static get _globals() {
        const globals = (global['cli-ux'] = global['cli-ux'] || {});
        globals.action = globals.action || {};
        return globals;
    }
}
exports.Config = Config;
