// cli-engine-heroku/lib/vars.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Vars = undefined;

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Vars {
  constructor(env) {
    this.env = env;
  }

  get host() {
    return this.env.HEROKU_HOST || 'heroku.com';
  }
  get apiUrl() {
    return this.host.startsWith('http') ? this.host : `https://api.${this.host}`;
  }
  get apiHost() {
    if (this.host.startsWith('http')) {
      const u = _url2.default.parse(this.host);
      if (u.host) return u.host;
    }
    return `api.${this.host}`;
  }
  get gitHost() {
    if (this.env.HEROKU_GIT_HOST) return this.env.HEROKU_GIT_HOST;
    if (this.host.startsWith('http')) {
      const u = _url2.default.parse(this.host);
      if (u.host) return u.host;
    }
    return this.host;
  }
  get httpGitHost() {
    if (this.env.HEROKU_GIT_HOST) return this.env.HEROKU_GIT_HOST;
    if (this.host.startsWith('http')) {
      const u = _url2.default.parse(this.host);
      if (u.host) return u.host;
    }
    return `git.${this.host}`;
  }

  get gitPrefixes() {
    return [`git@${this.gitHost}:`, `ssh://git@${this.gitHost}/`, `https://${this.httpGitHost}/`];
  }
}

exports.Vars = Vars;
exports.default = new Vars(process.env);
