// heroku-apps/lib/commands/labs/disable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _cliEngineHeroku = require('cli-engine-heroku');

class LabsDisable extends _cliEngineHeroku.Command {

  async run() {
    let feature = this.args.feature;
    let request;
    let target;

    try {
      await this.heroku.get(`/account/features/${feature}`);
      request = this.disableFeature(feature);
      target = (await this.heroku.get('/account')).body.email;
    } catch (err) {
      if (err.http.statusCode !== 404) throw err;
      // might be an app feature
      if (!this.app) throw err;
      await this.heroku.get(`/apps/${this.app}/features/${feature}`);
      request = this.disableFeature(feature, this.app);
      target = this.app;
    }

    this.out.action.start(`Disabling ${this.out.color.green(feature)} for ${this.out.color.cyan(target)}`);
    await request;
  }

  disableFeature(feature, app) {
    return this.heroku.patch(app ? `/apps/${app}/features/${feature}` : `/account/features/${feature}`, {
      body: { enabled: false }
    });
  }
}
exports.default = LabsDisable;
LabsDisable.topic = 'labs';
LabsDisable.command = 'disable';
LabsDisable.description = 'disables an experimental feature';
LabsDisable.args = [{ name: 'feature' }];
LabsDisable.flags = {
  app: _cliEngineHeroku.flags.app(),
  remote: _cliEngineHeroku.flags.remote()
};
